agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -197,52 +190,99 @@ public Function thenSort(@NonNull final Comparator comparator) {
     @SuppressWarnings("unchecked")
     public Object apply(@NonNull final Object input) {
       Object item = input;
-      for (final Function function : functions) {
-        item = function.apply(item);
-      }
-      return item;
-    }
-  }
-
-  private static final class FilterFunction<T> implements Function<List<T>, List<T>> {
-    @NonNull
-    private final Predicate filter;
+      boolean compilerList = false;
+      int i = 0;
+      final int length = functions.length;
+      while (i < length) {
+        final int type = (int) functions[i++];
+        final Object command = functions[i++];
+        switch (type) {
+          case LIMIT: {
+            final List list = (List) item;
+            final int limit = (int) command;
+            if (limit <= 0) {
+              item = emptyList();
+            } else if (list.size() < limit) {
+              item = input;
+            } else if (compilerList) {
+              while (list.size() > limit) {
+                list.remove(list.size() - 1);
+              }
+            } else {
+              final ArrayList newList = new ArrayList(limit);
+              for (int copied = 0; copied < limit; copied++) {
+                newList.add(list.get(copied));
+              }
+              item = newList;
+              compilerList = true;
+            }
+          }
+          break;
+          case FILTER: {
+            final List list = (List) item;
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else if (compilerList) {
+              final Predicate predicate = (Predicate) command;
```

Comment:
if predicate removes every other item, it will take O(N^2) to perform the filtering.

I think what should actually be here is:
1. go from the beginning to the end until predicate is true.
2. keep index of the firs item to remove (let's call it R)
3. continue iterate until predicate is false
4. if not at the end - copy element to R, and R++. If at the end go to 6.
5. go to 3.
6. remove elements from the end to R.